### PR TITLE
refactor : Relocate json schemas

### PR DESCRIPTION
**Pull request by antwigambrah**
## Proposed Commit Message

```
refactor: relocate json schemas 

Fixes GH-4688 
```

## Additional Context

fixes https://github.com/canonical/cloud-init/issues/4688

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [ ] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by catmsred** [comment]
Thank you for helping improve cloud-init!

A general question -- why did you decide on the schemas for the top level directory name as opposed to data as suugested in the bug report?

I think you may also need to update the schemas themselves, e.g. `"$id": "https://raw.githubusercontent.com/canonical/cloud-init/main/cloudinit/config/schemas/versions.schema.cloud-config.json"` to `"$id": "https://raw.githubusercontent.com/canonical/cloud-init/main/schemas/versions.schema.cloud-config.json"` although I haven't checked.

Can you provide some context on how you've tested your changes?

File: cloudinit/config/schema.py [hunk 1]
```diff
@@ -1429,7 +1430,7 @@ def load_doc(requested_modules: list) -> str:
 
 
 def get_schema_dir() -> str:
-    return os.path.join(os.path.dirname(os.path.abspath(__file__)), "schemas")
+    return os.path.join(Path(__file__).parent.parent.parent, "schemas")
```
Comment (TheRealFalcon) [inline cloudinit/config/schema.py:1433]: This unfortunately won't work. Only the `cloudinit` directory gets deployed as a unit, not the parent `cloud-init` directory. In order to accommodate this, we'll need to update `setup.py` to write the schema files to correct location on installation.

I would think somewhere like `/usr/lib/cloud-init/schemas/` would be most appropriate, but I don't think we have a definite directory for something like this yes.

@holmanb , thoughts?

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]

**Comment by antwigambrah** [comment]
> Thank you for helping improve cloud-init!
> 
> A general question -- why did you decide on the schemas for the top level directory name as opposed to data as suugested in the bug report?
> 
> I think you may also need to update the schemas themselves, e.g. `"$id": "https://raw.githubusercontent.com/canonical/cloud-init/main/cloudinit/config/schemas/versions.schema.cloud-config.json"` to `"$id": "https://raw.githubusercontent.com/canonical/cloud-init/main/schemas/versions.schema.cloud-config.json"` although I haven't checked.
> 
> Can you provide some context on how you've tested your changes?

For testing, I didn't add any additional test code but made sure   test_shema.py. Any additional kind of test you would like me run is welcome . Thanks for letting me know of the other location, will update it accordingly.

File: cloudinit/config/schema.py (hunk 1 above)
Comment (holmanb) [reply cloudinit/config/schema.py:1433]: > This unfortunately won't work. Only the cloudinit directory gets deployed as a unit, not the parent cloud-init directory. In order to accommodate this, we'll need to update setup.py to write the schema files to correct location on installation.

+1, agreed

> I would think somewhere like /usr/lib/cloud-init/schemas/ would be most appropriate, but I don't think we have a definite directory for something like this yes.

Agreed, although /usr/lib is for [libraries, shared objects, and internal executables](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s06.html). Maybe a cloud-init directory under [/usr/share](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s11.html) would be more appropriate?

Comment (blackboxsw) [reply cloudinit/config/schema.py:1433]: +1 on `/usr/share/cloud-init/schemas` as that aligns well with FHS for architecture independent data. We have some good examples of json schema files from other packages:
- vscode: /usr/share/code/resources
- /usr/share/lintian/data
- /usr/share/gnome-shell/extensions


Also note when we move forward on this approach we'll may need a one-liner followup PR against [schemastore updating the github URLs for the base schema](https://github.com/SchemaStore/schemastore/pull/2252/files#diff-1b8f038f5afb1158263a1fc83b9c0ca5a7438cceb90adad99d87168f70edc815R785) so that cloud-init schema can be auto-ingested by tools that automatically validate our user-data schema

Comment (antwigambrah) [reply cloudinit/config/schema.py:1433]: I can see in setup.py there is an existing /share/doc/cloud-init folder in which the schemas folder can also be created . Should it be maintained I should stick with moving the files to /usr/share/cloud-init/schemas .

**Comment by antwigambrah** [comment]
> Thank you for helping improve cloud-init!
> 
> A general question -- why did you decide on the schemas for the top level directory name as opposed to data as suugested in the bug report?
> 
> I think you may also need to update the schemas themselves, e.g. `"$id": "https://raw.githubusercontent.com/canonical/cloud-init/main/cloudinit/config/schemas/versions.schema.cloud-config.json"` to `"$id": "https://raw.githubusercontent.com/canonical/cloud-init/main/schemas/versions.schema.cloud-config.json"` although I haven't checked.
> 
> Can you provide some context on how you've tested your changes?

I have update the code based on suggestions made by @catmsred  @TheRealFalcon  and @blackboxsw  . I have added  test cases for the new file location and the ability to load from the location .

File: cloudinit/config/schema.py (hunk 1 above)
Comment (holmanb) [reply cloudinit/config/schema.py]: > I can see in setup.py there is an existing /usr/share/doc/cloud-init folder in which the schemas folder can also be created . Should it be maintained I should stick with moving the files to /usr/share/cloud-init/schemas .

Yes, I think that we agree that it should go there.

**Review by holmanb — CHANGES_REQUESTED** [review]
@antwigambrah sorry for the delayed response. Yes, lets put it in `/usr/shar/cloud-init/schemas/`. That aligns with FHS expectations.

**Comment by antwigambrah** [comment]
> @antwigambrah sorry for the delayed response. Yes, lets put it in `/usr/shar/cloud-init/schemas/`. That aligns with FHS expectations.

updated schema locations

**Comment by antwigambrah** [comment]
Hi @TheRealFalcon  any update on this pr

File: cloudinit/config/schema.py [hunk 2]
```diff
@@ -1553,8 +1553,12 @@ def load_doc(requested_modules: list) -> str:
     return docs
 
 
-def get_schema_dir() -> str:
-    return os.path.join(os.path.dirname(os.path.abspath(__file__)), "schemas")
+def get_schema_dir(schema_dir=None) -> str:
+
+     if schema_dir:
+        return os.path.join(schema_dir, "schemas")
```
Comment (holmanb) [inline cloudinit/config/schema.py:1559]: This path never actually gets called does it?

File: cloudinit/config/schema.py [hunk 3]
```diff
@@ -42,7 +42,7 @@
     load_text_file,
     write_file,
 )
-
```
Comment (holmanb) [inline cloudinit/config/schema.py:45]: unnecessary whitespace change

File: tests/unittests/config/test_schema.py [hunk 4]
```diff
@@ -128,6 +130,7 @@ def test_versioned_cloud_config_schema_is_valid_json(
         self, schema, error_msg
     ):
         schema_dir = get_schema_dir()
+
```
Comment (holmanb) [inline tests/unittests/config/test_schema.py:133]: unnecessary whitespace change

**Review by holmanb — CHANGES_REQUESTED** [review]
@antwigambrah Thanks for following up on this PR, and great work so far! I see that you've signed the CLA, but we need to add your github username to https://github.com/canonical/cloud-init/blob/main/tools/.github-cla-signers. Can you please include that in this PR?

This is generally looking good. I left a couple of nitpicks inline.

The issue that have with the current state of this PR is that, while unittests pass currently, I don't see a way for cloud-init to find the correct location on the filesystem to load these files when packaged for distros. I'm not sure about the best way to resolve this, but I suspect that it may involve a default path in `settings.py` so that distros can patch it if needed, however I'm not certain how best to make sure that when we want cloud-init to load the files from the relative codepath it will work while still working when installed. I'll try to come up with something, but if you have any ideas that would be appreciated!
